include/alpaka/queue/QueueGenericSyclBlocking.hpp

```diff
@@ -0,0 +1,207 @@
+
+        //! The SYCL blocking queue.
+        template<typename TDev>
+        class QueueGenericSyclBlocking
+        {
+        public:
+            QueueGenericSyclBlocking(TDev const& dev)
+                : m_dev{dev}
+                , m_queue{
+                      dev.m_context, // This is important. In SYCL a device can belong to multiple contexts.
+                      dev.m_device,
+                      {sycl::property::queue::enable_profiling{}, sycl::property::queue::in_order{}}}
+            {
+            }
+
+            auto operator==(QueueGenericSyclBlocking const& rhs) const -> bool
+            {
+                return (m_dev == rhs.m_dev) && (m_event == rhs.m_event);
+            }
+
+            auto operator!=(QueueGenericSyclBlocking const& rhs) const -> bool
+            {
+                return !operator==(rhs);
+            }
+
+            TDev m_dev; //!< The device this queue is bound to.
+            sycl::queue m_queue; //!< The underlying SYCL queue.
+            sycl::event m_event{}; //!< The last event in the dependency graph.
+            std::vector<sycl::event> m_dependencies = {}; //!< A list of events this queue should wait for.
+            std::shared_ptr<std::shared_mutex> mutex_ptr{std::make_shared<std::shared_mutex>()};
+        };
+    } // namespace experimental
+
+    namespace traits
+    {
+        //! The SYCL blocking queue device type trait specialization.
+        template<typename TDev>
+        struct DevType<experimental::QueueGenericSyclBlocking<TDev>>
+        {
+            using type = TDev;
+        };
+
+        //! The SYCL blocking queue device get trait specialization.
+        template<typename TDev>
+        struct GetDev<experimental::QueueGenericSyclBlocking<TDev>>
+        {
+            static auto getDev(experimental::QueueGenericSyclBlocking<TDev> const& queue)
+            {
+                std::shared_lock<std::shared_mutex> lock{*queue.mutex_ptr};
+                return queue.m_dev;
+            }
+        };
+
+        //! The SYCL blocking queue event type trait specialization.
+        template<typename TDev>
+        struct EventType<experimental::QueueGenericSyclBlocking<TDev>>
+        {
+            using type = experimental::EventGenericSycl<TDev>;
+        };
+
+        //! The SYCL blocking queue enqueue trait specialization.
+        template<typename TDev, typename TTask>
+        struct Enqueue<experimental::QueueGenericSyclBlocking<TDev>, TTask>
+        {
+            static auto enqueue(experimental::QueueGenericSyclBlocking<TDev>& queue, TTask const& task) -> void
+            {
+                if constexpr(experimental::detail::is_sycl_enqueueable<TTask>::value) // Device task
+                {
+                    std::scoped_lock<std::shared_mutex, std::shared_mutex, std::shared_mutex> lock{
+                        *queue.mutex_ptr,
+                        *queue.m_dev.mutex_ptr,
+                        task.pimpl->mutex};
```

Comment:
This *really* worries me.

Reply:
Are you worrying about deadlocks here?

Reply:
No, `scoped_lock` should take care of this. I am primarily worried about the design. Why do you need to aquire locks for 3 different objects to schedule a task? What are the implications of that? That means you cannot enqueue another kernel on a different thread to a different queue on the same device? Or enqueue the same task twice on two different threads/queues/devices. I am worried about design and performance implications.

Reply:
> Why do you need to aquire locks for 3 different objects to schedule a task?

Because we decided that our queues have to be thread-safe. And unfortunately alpaka's model of events and queues doesn't map very well to SYCL's way of handling events, queues and dependencies, so this is a (admittedly crude) work-around for the problems coming with this.

Reply:
As I said, we need a design review VC and discuss what this means exactly.